// @tensorflow-models/universal-sentence-encoder/dist/tokenizer/trie.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("../util");
var TrieNode = (function () {
    function TrieNode(key) {
        this.key = key;
        this.parent = null;
        this.children = {};
        this.end = false;
    }
    TrieNode.prototype.getWord = function () {
        var output = [];
        var node = this;
        while (node !== null) {
            if (node.key !== null) {
                output.unshift(node.key);
            }
            node = node.parent;
        }
        return [output, this.score, this.index];
    };
    return TrieNode;
}());
var Trie = (function () {
    function Trie() {
        this.root = new TrieNode(null);
    }
    Trie.prototype.findAllCommonPrefixes = function (ss, node, arr) {
        if (node.end) {
            var word = node.getWord();
            if (ss.slice(0, word[0].length).join('') === word[0].join('')) {
                arr.unshift(word);
            }
        }
        for (var child in node.children) {
            this.findAllCommonPrefixes(ss, node.children[child], arr);
        }
    };
    Trie.prototype.insert = function (word, score, index) {
        var node = this.root;
        var symbols = util_1.stringToChars(word);
        for (var i = 0; i < symbols.length; i++) {
            if (!node.children[symbols[i]]) {
                node.children[symbols[i]] = new TrieNode(symbols[i]);
                node.children[symbols[i]].parent = node;
            }
            node = node.children[symbols[i]];
            if (i === symbols.length - 1) {
                node.end = true;
                node.score = score;
                node.index = index;
            }
        }
    };
    Trie.prototype.commonPrefixSearch = function (ss) {
        var node = this.root.children[ss[0]];
        var output = [];
        if (node) {
            this.findAllCommonPrefixes(ss, node, output);
        }
        else {
            output.push([[ss[0]], 0, 0]);
        }
        return output;
    };
    return Trie;
}());
exports.Trie = Trie;
